src/enzyme_ad/jax/Passes/EnzymeHLOOpt.cpp

```diff
@@ -15485,6 +15485,170 @@ struct SliceWrap final : OpRewritePattern<enzymexla::WrapOp> {
+    int64_t extendDim = extendOp.getDimension();
+    int64_t lhs = extendOp.getLhs();
+    int64_t rhs = extendOp.getRhs();
+
+    // Get permutation array
+    SmallVector<int64_t> permutation;
+    for (auto val : op.getPermutation()) {
+      permutation.push_back(static_cast<int64_t>(val));
+    }
+
+    // The new extend dimension will be the permuted dimension
+    int64_t newExtendDim = -1;
+    for (size_t i = 0; i < permutation.size(); ++i) {
+      if (permutation[i] == extendDim) {
+        newExtendDim = i;
+        break;
+      }
+    }
+
+    if (newExtendDim == -1)
+      return failure();
+
+    // First transpose the extend's operand
+    auto newTranspose = rewriter.create<stablehlo::TransposeOp>(
+        op.getLoc(), extendOp.getOperand(), op.getPermutation());
+
+    // Then create a new extend operation on the transposed data
+    auto newExtendType = op.getType();
+    auto newExtendOp = rewriter.create<enzymexla::ExtendOp>(
+        op.getLoc(), newTranspose.getResult(), lhs, rhs, newExtendDim);
+
+    // Replace the original op with the new extend operation
+    rewriter.replaceOp(op, newExtendOp);
+    return success();
+  }
+};
+
+// transpose(rotate) -> rotate(transpose)
+struct TransposeRotate final : OpRewritePattern<mlir::stablehlo::TransposeOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(mlir::stablehlo::TransposeOp op,
+                                PatternRewriter &rewriter) const override {
+    auto type = dyn_cast<RankedTensorType>(op.getType());
+    if (!type)
+      return failure();
+
+    // Check if the operand is a rotate operation
+    auto rotateOp = op.getOperand().getDefiningOp<enzymexla::RotateOp>();
+    if (!rotateOp)
+      return failure();
+
+    // Get rotate operation parameters
+    int64_t rotateDim = rotateOp.getDimension();
+    int64_t amount = rotateOp.getAmount();
+
+    // Get permutation array
+    SmallVector<int64_t> permutation;
+    for (auto val : op.getPermutation()) {
+      permutation.push_back(static_cast<int64_t>(val));
+    }
+
+    // The new rotate dimension will be the permuted dimension
+    int64_t newRotateDim = -1;
+    for (size_t i = 0; i < permutation.size(); ++i) {
+      if (permutation[i] == rotateDim) {
+        newRotateDim = i;
+        break;
+      }
+    }
+
+    if (newRotateDim == -1)
+      return failure();
+
+    // First transpose the rotate's operand
+    auto newTranspose = rewriter.create<stablehlo::TransposeOp>(
+        op.getLoc(), rotateOp.getOperand(), op.getPermutation());
+
+    // Then create a new rotate operation on the transposed data
+    auto newRotateType = op.getType();
+    auto newRotateOp = rewriter.create<enzymexla::RotateOp>(
+        op.getLoc(), newTranspose.getResult(), amount, newRotateDim);
+
+    // Apply shardings if present in the original ops
+    if (auto shard = sdy::getShardingPerValue(op)) {
+      sdy::setShardings(newRotateOp, shard);
+    }
```

Comment:
I'm not sure whether the handling of sharding is correct here. I assume I need to add checks like this for the other transforms as well.

Reply:
don't worry about sharding stuff here, that's only required for recognize/lower since we want to be able to use it post shard